nibiru/websocket.py

```diff
@@ -0,0 +1,106 @@
+import json
+import threading
+import time
+from multiprocessing import Queue
+from typing import List
+
+from websocket import WebSocketApp
+
+from nibiru import Network
+from nibiru.event_specs import EventCaptured, EventType
+
+ERROR_TIMEOUT_SLEEP = 3
+
+
+class NibiruWebsocket:
+    queue: Queue = None
+    captured_events_type: List[List[str]]
+
+    def __init__(
+        self,
+        network: Network,
+        captured_events_type: List[EventType] = [],
+    ):
+        """
+        The nibiru listener provides an interface to easily connect and handle subscription to the events of a nibiru
+        chain.
+        """
+
+        self.websocket_url = network.websocket_endpoint
+        self.captured_events_type: dict = {
+            captured_event.value: captured_event
+            for captured_event in captured_events_type
+        }
+        self.queue = Queue()
+
+    def start(self):
+        """
+        Start the websocket and fill the queue with events.
+        """
+
+        self._ws = WebSocketApp(
+            self.websocket_url,
+            on_open=self._on_open,
+            on_message=self._on_message,
+            on_error=self._on_error,
+        )
+        threading.Thread(
+            target=self._ws.run_forever,
+            daemon=True,
+            name=f"Nibiru websocket @ {self.websocket_url}",
+        ).start()
+
+    def _on_open(self, _: WebSocketApp):
+        print("WebSocket starting")
+        self._subscribe()
+
+    def _on_error(self, ws: WebSocketApp, error: Exception):
+        print(f"Closing websocket, error {error}")
+        ws.close()
+        time.sleep(ERROR_TIMEOUT_SLEEP)
+        ws.run_forever()
+
+    def _on_message(self, _: WebSocketApp, message: str):
+        """
+        Parse the message and filter through them using the captured event type.
+        Put these filtered event in the queue.
+
+        Args:
+            _ (WebSocketApp): No idea what this is
+            message (str): The message in a utf-8 data received from the server
+        """
+        message_time = time.time_ns()
```

Comment:
We should use block time, not just current time. Although as far as I see - none of the Tx result includes a block time. So maybe we just skip this and add separate method for retrieving block time by block number somewhere else?

Reply:
I'll need to add back the rpc url into the network object for this. I'll do this in another PR, where we also check for the version of the chain with the rpc